### PR TITLE
Treat dag-defined access_control as authoritative if defined

### LINKED ISSUE #32839: DAG-level permissions set in Web UI disappear from roles on DAG sync
### Apache Airflow version

2.6.3

### What happened

Versions: 2.6.2, 2.6.3, main

PR [#30340](https://github.com/apache/airflow/pull/30340) introduced a bug that happens whenever a DAG gets updated or a new DAG is added

**Potential fix:** Adding the code that was removed in PR [#30340](https://github.com/apache/airflow/pull/30340) back to `airflow/models/dagbag.py` fixes the issue. I've tried it on the current main branch using Breeze.

### What you think should happen instead

Permissions set in Web UI stay whenever a DAG sync happens

### How to reproduce

1. Download `docker-compose.yaml`:
```bash
curl -LfO 'https://airflow.apache.org/docs/apache-airflow/2.6.2/docker-compose.yaml'
```
2. Create dirs and set the right Airflow user:
```bash
mkdir -p ./dags ./logs ./plugins ./config && \
echo -e "AIRFLOW_UID=$(id -u)" > .env
```
4. Add `test_dag.py` to ./dags:
```python
import datetime

import pendulum

from airflow import DAG
from airflow.operators.bash import BashOperator

with DAG(
    dag_id="test",
    schedule="0 0 * * *",
    start_date=pendulum.datetime(2021, 1, 1, tz="UTC"),
    catchup=False,
    dagrun_timeout=datetime.timedelta(minutes=60),
) as dag:
    test = BashOperator(
        task_id="test",
        bash_command="echo 1",
    )
if __name__ == "__main__":
    dag.test()
```
5. Run docker compose: `docker compose up`
6. Create role in Web UI: Security > List Roles > Add a new record:
	Name: test
	Permissions: `can read on DAG:test`
7. Update `test_dag.py`: change `bash_command="echo 1"` to `bash_command="echo 2"`
8. Check test role's permissions: `can read on DAG:test` will be removed

Another option is to add a new dag instead of changing the existing one:
6. Add another dag to ./dags, code doesn't matter
7. Restart scheduler: `docker restart [scheduler container name]`
9. Check test role's permissions: `can read on DAG:test` will be removed

### Operating System

Ubuntu 22.04.1 LTS

### Versions of Apache Airflow Providers

_No response_

### Deployment

Docker-Compose

### Deployment details

Docker 24.0.2

### Anything else

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by SamWheating**
closes: https://github.com/apache/airflow/issues/32839

(Copying my comment from the original issue)

We have two places to define permissions (manually via the UI/CLI and in code via `DAG.access_control`) and there's no stance on which one takes precedence, which leads to all sorts of unexpected behaviour (mostly around [not cleaning up permissions enough](https://github.com/apache/airflow/issues/25149), or [cleaning up permissions too aggressively](https://github.com/apache/airflow/issues/32839))

I would propose making the access control for a given DAG authoritative as long as it is explicitly set. This way, if a DAG was updated, removing `access_control` from a DAG would not trigger a cleanup, but setting `access_control={}` would.

So we have three cases. A DAG defined like this:
```python
with DAG(
    dag_id="test",
    schedule="0 0 * * *",
    start_date=pendulum.datetime(2021, 1, 1, tz="UTC"),
    access_control={'role1': {'can_read'}, 'role2': {'can_read', 'can_edit', 'can_delete'}}
) as dag:
```
Would overwrite any existing dag-level permissions for this DAG, ensuring that the roles listed here are the only ones with permissions specific to this DAG.

While a DAG with explicitly empty access_contol, like this:
```python
with DAG(
    dag_id="test",
    schedule="0 0 * * *",
    start_date=pendulum.datetime(2021, 1, 1, tz="UTC"),
    access_control={}
) as dag:
```
Would explicitly clear out all DAG-level permissions for this role.

While a DAG without any access_control defined:
```python
with DAG(
    dag_id="test",
    schedule="0 0 * * *",
    start_date=pendulum.datetime(2021, 1, 1, tz="UTC"),
) as dag:
```
Would just defer to the existing DAG-level roles in the database.

This presents its own hazards around dangling and stray permissions, but I think that any approach here is going to have some sort of downside. The important part is that the approach to DAG-level access is clearly-defined and well documented (neither of which is true at the moment).

#### TODO (before merging): 

I'll do these things as a part of this PR, but I want to get early feedback on the approach before I spend too much time here.

 - [ ] Add documentation around the precedence of different DAG-level access controls.
 - [ ] Update and add tests

### CODE REVIEW

File: airflow/www/security.py [hunk 1]
```diff
@@ -681,26 +681,12 @@ def sync_perm_for_dag(
...
-                        "Revoking '%s' on DAG '%s' for role '%s'",
-                        perm.action,
-                        dag_resource_name,
-                        role.name,
-                    )
-                    self.remove_permission_from_role(role, perm)
-
-        if access_control:
+        
+        if access_control is not None:
```
Comment (SamWheating) [inline airflow/www/security.py:685]: This is the most important change in this PR - `if access control` will proceed into the following block if and only if access_control is a non-empty dict, whereas `if access control is not None` will proceed given any dict, including an empty one.

**Comment by potiuk** [comment]
Yes. I agree with the reasoning. 

REbasing now to incorporafe to hace US-ntity

**Review by potiuk — APPROVED** [review]
LGTM.  Needs static code check /docs fixing. I like the docs and explanation and finally a bit more clarity on what the users should expect.

File: docs/apache-airflow/security/access-control.rst [hunk 2]
```diff
@@ -229,3 +229,39 @@ List Plugins                           Plugins.can_read
...
+    :end-before: [END howto_fine_grained_access_control]
+
+ - Setting `access_control={}` will wipe any existing DAG-level permissions for a given DAG from the DB.
+
+.. exampleinclude:: /../../airflow/example_dags/example_dag_level_access_control.py
+    :language: python
+    :start-after: [START howto_no_fine_grained_access_control]
+    :end-before: [END howto_no_fine_grained_access_control]
+
+ - Removing the access_control block from a DAG altogether (or setting it to `None`) can leave dangling permissions.
```
Comment (potiuk) [inline docs/apache-airflow/security/access-control.rst]: Yep. This is a side effect that we can accept.

One small comment. Maybe a good idea would be to also accompany it with a new CLI command `airflow dags clean-permissions` and a new API call (eventually) to do it. Those can be separate PRs, but having them explicitly spelled out as way to clean permissions on a DAG that previously had permissions is a way to workaround the problem with dangling permissions. 

I see that as an edge case, with this change there is already a way to clean permissions by setting access_control to `{}` - but I can see a situation that somoene would like to recover from the historical changes.

**Review by huymq1710 — APPROVED** [review]
LGTM
Thank you very much

**Comment by potiuk** [comment]
Some static checks :(, Recommend installing pre-commit. Also `breeze statich-checks --only-my-changes` is helpful.

**Comment by SamWheating** [comment]
> breeze static-checks --only-my-changes

Good to know, I was just running ruff + black checks but it looks like there's a few more failing 👀 I'll use this command in the future.

**Comment by potiuk** [comment]
> > breeze static-checks --only-my-changes
> 
> Good to know, I was just running ruff + black checks but it looks like there's a few more failing 👀 I'll use this command in the future.

Hihly recommend just `pre-commit install` I generally almost never run any static checks manually for airflow. It happens all automagically with pre-commits, so I just don't think about them and fix them when I commit stuff.

**Comment by potiuk** [comment]
Still small thing in code blocks :(

**Comment by potiuk** [comment]
🎉

**Pull request merged**